### PR TITLE
add support for custom project domain in user identity.

### PR DESCRIPTION
**Pull request by acontes**
The changes add the ability to connect to openstack instances having
the project's domain name different from the domain name.
It introduces an optional format for the user identity field while
remaining compatible with previous configurations of the
plugin.

### CODE REVIEW

**Comment by olivergondza** [comment]
I must admit I am confused by all the OS domains. So the `domain` we already have is for the domain _user_ belongs to and the `project-domain` is the one the _project_ belongs to and they can be different in general? With the latter being optional and used just to disambiguate colliding project names between project domains?

**Comment by acontes** [comment]
In Openstack, both projects (tenant) and users must belong to a domain. 
In a 'simple' configuration, projects and users are created within the same domain but nothing enforces it. For the sake of some security rules or different authentication mechanisms, projects and users can be created in different domains. 

The fact that project and users must be linked to a domain is documented in the [Openstack documentation](https://docs.openstack.org/mitaka/install-guide-obs/keystone-users.html). Nothing specifies that they have to share the same domain.

In the code of the pull-request, I have decided to make the project's domain optional to maintain compatibility with the current implementation.

File: src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-identity.html [hunk 1]
```diff
@@ -3,6 +3,6 @@
   The format required depends on the authentication method used by the OpenStack system.
   <ul>
   <li>For v2 authentication use syntax <tt>TENANT_NAME:USER_NAME</tt></li>
-  <li>For v3 authentication use syntax <tt>PROJECT_NAME:USER_NAME:DOMAIN_NAME</tt></li>
+  <li>For v3 authentication use syntax <tt>[PROJECT_DOMAIN\]PROJECT_NAME:USER_NAME:DOMAIN_NAME</tt> PROJECT_DOMAIN is optional, use only when DOMAIN_NAME and PROJECT_DOMAIN differs</li>
```
Comment (Guillaumichaud) [inline src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-identity.html]: differ

**Comment by acontes** [comment]
typo fixed, thank you @Guillaumichaud for the review

**Comment by olivergondza** [comment]
Right, that was my understanding too. Is there a specific reason to use backslash as a separator and not a `:`? Also, for user we use `user:domain` ordering and for project it would be `domain:project`. How about `PROJECT_NAME:PROJECT_DOMAIN:USER_NAME:DOMAIN_NAME`?

**Comment by acontes** [comment]
fine by me, I will update the PR accordingly.

**Comment by pjdarton** [comment]
I think that all this foo:bar:optional stuff is just highlighting some tech-debt that we really ought to stuff all this in its own multi-field class.  Not as part of this PR but I think we ought to do it...

When it becomes difficult to explain everything in the online help text, that indicates that it's overdue for being refactored ;-)

**Comment by olivergondza** [comment]
@pjdarton, certainly. Though, I hesitate to do it because of all the ajax method signatures getting nasty. IIRC, one of yours that would be painfully affected.

**Comment by pjdarton** [comment]
Yes, the doCheck.../doFill... methods will get a lot more arguments _but_ it's only us that'll see that - the end user will get a much cleaner experience.
It's certainly not something to be done while we've got some large PRs awaiting merger ;-)

**Comment by acontes** [comment]
the updated version simplifies the identity string (as so the documentation) as the domain name for the project is now mandatory.
The migration to the new v3 schema is performed in the readResolve method.

**Comment by acontes** [comment]
Hello there, any comment/remark on the latest version ?

**Comment by acontes** [comment]
PR updated to the latest version of master.

**Comment by malice00** [comment]
I just tried these changes and they solved my authentication issues! I hope this can get merged and released soon?

**Comment by olivergondza** [comment]
I agree with @pjdarton, that this is best refactored into (at least) 2 fields for project and username.

**Comment by Guillaumichaud** [comment]
@acontes Was the "Unresolveable build extension" error expected? :)

**Comment by acontes** [comment]
rather than just adding a second field, I updated the code to use Jenkins' credentials and create two implementations ( one for v2 and one for v3 ). Adding a second field has the same impact on the API but is far less extensible.
As for the previous version of the pull-request, the migration from the former schema is performed in the readResolve method.

File: src/main/java/jenkins/plugins/openstack/compute/internal/Openstack.java [hunk 2]
```diff
@@ -738,21 +721,25 @@ private static void debug(@Nonnull String msg, @Nonnull String... args) {
...
         /**
          * Instantiate Openstack client.
          */
         public static @Nonnull Openstack get(
-                @Nonnull final String endPointUrl, @Nonnull final String identity, @Nonnull final String credential, @CheckForNull final String region
+                @CheckForNull final String endPointUrl, @CheckForNull final OpenstackCredential auth, @CheckForNull final String region
         ) throws FormValidation {
-            final String fingerprint = Util.getDigestOf(endPointUrl + '\n' + identity + '\n' + credential + '\n' + region);
+            if (auth == null) {
+                throw new RuntimeException("authentication is null");
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/internal/Openstack.java]: Why not keeping the parameters `@Nonnull` then?

File: src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-credentialId.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<div>
+    The credential to start the machine.
```
Comment (olivergondza) [inline src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-credentialId.html]: This is the credential to talk to openstack.

File: src/main/resources/jenkins/plugins/openstack/compute/auth/AbstractOpenstackCredential/config.jelly [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define"
+	xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+</j:jelly>
```
Comment (olivergondza) [inline src/main/resources/jenkins/plugins/openstack/compute/auth/AbstractOpenstackCredential/config.jelly]: Do we need a `config.jelly` for the abstract class if it is overridden by all implementations?

File: src/main/resources/jenkins/plugins/openstack/compute/auth/OpenstackCredentialv2/help-tenant.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<div>
+    The tenant associated to the user to start the machine.
```
Comment (olivergondza) [inline src/main/resources/jenkins/plugins/openstack/compute/auth/OpenstackCredentialv2/help-tenant.html]: I find the "to start the machine" part in auth helps a bit confusing - it is needed for any interaction with openstack. I guess we can leave it out.

File: src/main/resources/jenkins/plugins/openstack/compute/auth/OpenstackCredentialv3/help-username.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div>
+    User name of the user to start the machine.
```
Comment (olivergondza) [inline src/main/resources/jenkins/plugins/openstack/compute/auth/OpenstackCredentialv3/help-username.html]: "to start the machine" is extremely confusing here.

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java [hunk 7]
```diff
@@ -60,10 +64,14 @@
 public class JCloudsCloud extends Cloud implements SlaveOptions.Holder {
 
     private static final Logger LOGGER = Logger.getLogger(JCloudsCloud.class.getName());
+    public String credentialId;
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: `private`

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java [hunk 8]
```diff
@@ -60,10 +64,14 @@
...
 
     public final @Nonnull String endPointUrl;
-    public final @Nonnull String identity;
-    public final @Nonnull Secret credential;
+
+    public /*final*/ @Nonnull String identity;
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: This one should be `@Deprecated` as well, right?

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java [hunk 9]
```diff
@@ -144,6 +155,33 @@ private Object readResolve() {
...
+                String domain = id[2];
+                migratedOpenstackCredential = new OpenstackCredentialv3(CredentialsScope.SYSTEM,null,null,username,domain,project,domain,credential);
+            }
+            if (migratedOpenstackCredential != null) {
+                credentialId = migratedOpenstackCredential.getId();
+                try {
+                    OpenstackCredentials.add(migratedOpenstackCredential);
+                    OpenstackCredentials.save();
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: Send to logger, instead.

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java (hunk 8 above)
Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: ... and transient.

**Comment by olivergondza** [comment]
Merged, thanks!

**Comment by twillouer** [comment]
:+1:

**Pull request merged**